Cosmos DB: Add schema inference

ealsur opened the pull request:
This PR enables schema inference on the connector. Inference is achieved by sampling a number of documents from the source collection and used them to define the schema.

**Schema inference kicks in if the user does not explicitly define the schema when performing a read**

## New configurations

* `spark.cosmos.read.inferSchemaSamplingSize`: Defines the sampling size (default 1000)
* `spark.cosmos.read.inferSchemaEnabled`: Disables or enables schema inference (default disabled)

If `spark.cosmos.read.inferSchemaEnabled` is disabled ("false") and the user did not specify a schema, a simple default schema is generated which contains:

* `_ts` field for the timestamp value (LongType)
* `id` field (StringType)
* `_rawBody` field, which when ran through the RowConverter, contains the raw json for the whole document (StringType)

This is useful in cases where the raw Json is expected to be sent somewhere else.

If `spark.cosmos.read.inferSchemaEnabled` is enabled ("true") and the user did not specify a schema, then the documents from the source will be sampled in order to infer one.

Schema is defined as a list of unique fields and types used across all sampled documents. For example, if the source contains:

```
{
    "field1": "aString"
},
{
    "field2": 10
}
```

Then the schema will have both fields, one as a StringType, the other as an Integer type. When the Rows are generated, each one will have a null value on the missing attribute.

## Tests

Added Specs for the new inference code, some new tests for the Converter to cover raw json conversion, and E2E with inferSchemaEnabled = false

Closes #18152

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:
```diff
@@ -0,0 +1,116 @@
+    def inferSchema(inferredItems : Seq[ObjectNode]): StructType = {
+        if (inferredItems.isEmpty){
+            // No documents to infer from
+            StructType(Seq())
```
Should we return defaultSchemaForInferenceDisabled  instead?

ealsur replied:
I thought so but if you think about it, the collection is empty, so there will be no rows that will be read, the schema won't apply to anything, so the schema won't be used.

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:
```diff
@@ -0,0 +1,116 @@
+        }
+
+        // Create a unique map of all distinct properties from documents
+        // If 2 documents contain the same property name but different type, the last scanned one would define the final
```
Would it make sense to fall back to String in this case? We can pretty much convert all types into string, correct?

ealsur replied:
Done

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:
```diff
@@ -0,0 +1,116 @@
+
+            val queryOptions = new CosmosQueryRequestOptions()
+            queryOptions.setMaxBufferedItemCount(cosmosReadConfig.inferSchemaSamplingSize)
+            val queryText = s"select TOP ${cosmosReadConfig.inferSchemaSamplingSize} * from c"
```
Not blocking - just something to discuss I want to track here: 
This approach will make it very hard for customers to use schema inference if they store documents with multiple schemas in one container - like one common pattern for IoT scenarios is that for each device you store a summary (basically the digital twin with the "materialized view" of this device you need most commonly for the app/UI) and the raw events (usually with TTL of a couple of days) in the same container (to be able to insert event and update view in an atomic transaction). So I am wondering whether we can either pipe some of the filtering/query from Spark into the schema inference here (not sure whether you even have access to the query context yet) or whether we should allow to specify filtering conditions/query to be used for schema inference in the config options?

ealsur replied:
I checked other Spark integrations and they follow a similar approach (sampling documents), the thing is, with a NoSQL store, it's is rather complicated to infer the schema, like you said, there could be documents with varied schema. We could potentially add a config to inject a custom query?

FabianMeiswinkel replied:
Yes- I like the approach of having a separate optional config property to define the sampling query. Let's discuss with Mo - would be separate PR anyway.

moderakh replied:
I think adding a custom Query will be very helpful. yes. can go in a separate PR though.

ealsur replied:
Will do it in a following PR

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala:
```diff
@@ -0,0 +1,116 @@
+                case None => NullType
+                }
+            case _ =>
+                this.logError(s"Unsupported document node conversion [${jsonNode.getNodeType}]")
```
Info - maybe warning.

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:
```diff
@@ -167,8 +167,10 @@ object CosmosRowConverter
     private def convertStructToSparkDataType(schema: StructType, objectNode: ObjectNode) : Seq[Any] =
         schema.fields.map {
+            case StructField(CosmosTableSchemaInferer.RAW_JSON_BODY_ATTRIBUTE_NAME, StringType, _, _) =>
+                objectNode.toString
```
The opposite is missing - basically special-casing if the raw json and the required system properties for id/pk exist to use those on write

ealsur replied:
Makes sense

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:
```diff
@@ -92,6 +92,29 @@ object CosmosContainerConfig {
+        val samplingSize = CosmosConfigEntry.parse(cfg, inferSchemaSamplingSize)
+        val enabled = CosmosConfigEntry.parse(cfg, inferSchemaEnabled)
+
+        CosmosReadConfig(samplingSize.getOrElse(DefaultSampleSize), enabled.getOrElse(true))
```
I think enabled should have a default value of "false" - and I also think you still need to wire up the CosmosItemDataSource.inferSchema to return the same schema

I think it would be risky to apply an effective schema to the Table created by the TableProvider with a different schema than returned from the TableProvider.inferSchema call. For writes we can basically tell Spark to not validate the output of inferSchema with the ACCEPT_ANY_SCHEMA capability (which is fine - as long as you return the default schema when the spark.cosmos.read.inferSchemaEnabled" property isn't set this would simply work for writes. Worst case when customers set the property even on write it will give a small perf hit. By disabling it by default on reads you also only add the perf hit if explicitly requested.
My concern is that the TableProvider.inferschema is most likely called early by the catalyst infrastructure to generate the query plan - so if we return inconsistent schema it would skip possible optimizations and even cause failures on more complicated queries.

ealsur replied:
Defaulted to false

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala:
```diff
@@ -70,18 +71,12 @@ class CosmosTable(val transforms: Array[Transform],
   }
 
   override def capabilities(): util.Set[TableCapability] = Set(
+    TableCapability.ACCEPT_ANY_SCHEMA,
```
Please add a brief comment why this ended-up being needed .

ealsur replied:
Done

FabianMeiswinkel reviewed (changes requested):
Thanks Matias - looks great. Main issues that need to be addressed IMO are "wiring up TableProvider.inferSchema" and adding support to mpap raw json back to ObjectNode on the write path

FabianMeiswinkel commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTable.scala:
```diff
@@ -76,6 +76,9 @@ class CosmosTable(val transforms: Array[Transform],
   override def capabilities(): util.Set[TableCapability] = Set(
+    // ACCEPT_ANY_SCHEMA is needed because of this bug https://github.com/apache/spark/pull/30273
+    // It was fixed in Spark 3.1.0 but Databricks currently only supports 3.0.1
+    // When Databricks supports 3.1.X, we could remove it and move to a newer Spark version
```
Only in new major version - changing capabilities is a breaking change. Leave the first two lines and remove this one and we deal with deciding whether we want to take the new major version later.

ealsur replied:
Done

FabianMeiswinkel reviewed (changes requested):
Thanks Matias - LGTM

FabianMeiswinkel reviewed (approved):
LGTM

moderakh commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:
```diff
@@ -92,6 +92,29 @@ object CosmosContainerConfig {
 
+case class CosmosReadConfig(inferSchemaSamplingSize: Int, inferSchemaEnabled: Boolean)
+
+object CosmosReadConfig {
```
```suggestion
object CosmosSchemaInferenceConfig {
```

ealsur replied:
Done

moderakh commented on sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:
```diff
@@ -92,6 +92,29 @@ object CosmosContainerConfig {
   }
 }
 
+case class CosmosReadConfig(inferSchemaSamplingSize: Int, inferSchemaEnabled: Boolean)
```
renaming
```suggestion
case class CosmosSchemaInferenceConfig(inferSchemaSamplingSize: Int, inferSchemaEnabled: Boolean)
```

ealsur replied:
Done

moderakh reviewed (commented):
great work.

Could you please also add a basic end to end test for utilizing schema inference in the following files:

- SparkE2EQuerySpec
- SparkE2EWriteSpec

ealsur commented:
> great work.
> 
> Could you please also add a basic end to end test for utilizing schema inference in the following files:
> 
> * SparkE2EQuerySpec
> * SparkE2EWriteSpec

Thanks @moderakh, I addressed your rename comments and also added more tests, to SparkE2EQuerySpec. SparkE2EWriteSpec is not affected.

moderakh reviewed (approved):
LGTM.

Thank you Matias great work.
